### DISCUSSION TITLE
I would like to be able to serialize strings containing the null character

### DISCUSSION THREAD

**Discussion by olowo726** [zeux/pugixml#517] [Q&A]
First, if this is possible but I haven't found out how, please let me know how.

I need to serialize a C array of chars, say char buf[10], which might contain the null character '\0'. The format establishes, standardized and governed by a rather major standardization organization so it's not something you change. My issue is that as soon as a null character is encountered pugi aborts. So, what I would like the output to contain is 

`<VT xml:space="preserve">&#0;&#48;</VT>`

for the string "\08"

Can this be accomplished?

**Comment by zeux** [comment]
You can accomplish this by escaping the data yourself - storing `&#0;` in the string data you store in the tree, and use `format_no_escapes` flag when writing the data to a file. Of course, this would require escaping all data in the same fashion.

When parsing files like this, you'd need to also disable unescaping - using `parse_default & ~parse_escapes` option mask. Similarly, this would require unescaping the data during processing in your application.

Note that the resulting file is not valid per XML specification, as any characters escaped with `&#` must match https://www.w3.org/TR/xml/#NT-Char - but I understand that you can't change this.

**Reply by olowo726** [reply]
Interesting, I didn't knew about the range limitation for NT-char.

Today we have `xml_text& xml_text::operator=(const char_t* rhs);`, I guess that I'm looking for an API which allows me to pass both a pointer to the characters and length. But I understand this would also have implications for the API for how to retrieve the text when parsing a document (I also need to be able to read such XML documents)

**Reply by zeux** [reply]
The tree doesn't store lengths explicitly, instead relying on null termination of the string data. This is made possible by the aforementioned restriction on valid XML documents. So handling escaping in the application layer is the only available route right now.
